### ISSUE TITLE
Allow the uninstallation of NVDA to remove user configuration

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#6869]
Since add-ons reside in the user configuration and can take up a lot of space (especially synthesizer voices), it would increase user friendliness if the uninstallation would offer a possibility to remove the user configuration.

In my opinion, the best implementation would be to add user config removal to the installer module. After that, nvda_slave can be changed in a way that the uninstaller can request installer.unregisterInstallation to also remove the user configuration. The actual check box for user input should be added to the nsis part.

When #6822 gets added, should this option remove the central configuration as well?

**Comment by jcsteh** [comment]
Doing this is potentially misleading. It will (and can) only remove the
configuration for the current user, not any other users on the system.
Someone can thus be fooled into thinking it will remove all user config
when in fact it only removes config for a single user. I guess there's
precedent for this: if installing from portable, the installer can copy the
portable config for the current user. I think I made a similar argument
against that one too. :) The only difference here is that this will be seen
a lot more often; it's probably not that common to install from a portable.

@MichaelDCurran, @QChristensen, @feerrenrut, thoughts? I'm not a massive
fan of this idea for the above reason, but I'm not going to fight it if one
or all of you champions the cause. :)

**Comment by leonardder** [comment]
> Someone can thus be fooled into thinking it will remove all user 
config when in fact it only removes config for a single user.

I guess that can be accounted for by explicitly stating user 
configuration in the check box label. Another thing I missed though, is 
that we do not have a way to translate this string. That would mean 
nvda_slave should pop up a question, and personally I'm not a fan of 
that one.

**Comment by Qchristensen** [comment]
The main use case I can think of, is where a user is uninstalling and
reinstalling to try and fix something they've messed up in the
configuration.  I know that's not the best way to fix it, but this came up
just recently in the user list, so it happens.  I think if we went this
way, we'd need to clearly indicate that it only removes the config for the
current user.


--
22 Point

Check out our first app, RapiTap! - Tap targets fast & avoid decoys:
Adrenaline pumping, Challenging, Accessible!
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitap
Free trial:
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitapfree

Web: http://www.22point.com.au
Facebook: https://www.facebook.com/22Point
Twitter: https://twitter.com/22PointApps

**Comment by feerrenrut** [comment]
Can we clarify the problem to solve here? Is it to remove addons so that they do not take up lots of space? I think its a little confusing to ask the user to remove the config when the goal is to free up space taken by addons. Perhaps the user wants to keep the config but free up space (so that when the addon is reinstalled it does not need to be reconfigured)

**Comment by leonardder** [comment]
The major part of this problem is, that there is NVDA data left on the 
system whenever NVDA is removed. This is particularly noticeable in case 
of large add-ons.

**Comment by Brian1Gaff** [comment]
Yes you do certainly need some form of clean out if you run snaps a lot and 
want to install, say master instead of  next, as it complains you are 
downdating and its not a good thing to do, so then you need to go around 
removing stuff manually while running a portable copy of nvda which on 
windows 10 can be a real pain in the neck due to the way so much fails to 
work correctly due to it not being installed for the user  using the 
machine!

Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal email to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by Qchristensen** [comment]
On the use case of someone reinstalling NVDA to try to fix a configuration problem (since I just saw it come up again), I wonder is it worth having an option while installing to keep the previous configuration (only displayed where one is found).

**Comment by Adriani90** [comment]
I agree there should be at least a confirmation dialog when uninstalling NVDA which asks if the user wants to also remove all the settings, drivers and addons. Lots of beginners have some strange issues and we must explain everytime how to find the user config directory in %appdata% and how to remove it and then reinstall NVDA.

**Comment by Qchristensen** [comment]
Reading through this issue again, I figured I would create the issue that I suggested in my comment, so I have done that: https://github.com/nvaccess/nvda/issues/9596

**Comment by dingpengyu** [comment]
Can the current problem be solved?
thank

**Comment by feerrenrut** [comment]
As @jcsteh pointed out, even if uninstalling for this user, other users may still have addon data in their profiles. There isn't a solution to solve this yet. On the other hand, this might be making perfect the enemy of good. I would guess that the majority of systems are single user.

On the grander scheme, I'd like to see add-ons stored somewhere shared like app-data rather than the user profile. Then add-ons can be shared by users. This is a separate idea, that we don't need to discuss here, but would alleviate this issue.

**Comment by CyrilleB79** [comment]
@feerrenrut  or @jcsteh, can you clarify the following that I do not understand?

1. Why can only the current user's configuration be removed during uninstallation process? I imagine that the uninstaller should request privilege elevation and thus should be granted admin rights. Thus it should be able to remove all the folders such as:
`C:\Users\*\AppData\Roaming\nvda`
and also
`C:\Users\*\AppData\Local\nvda`

2. Many applications (such as Jaws, Matlab, but maybe also browsers, Notepad++) ask if user config should also be removed or not when uninstalled. I have always imagined that checking the "Also remove user data" caused data removal for all users. But I have never checked. Do you know more on this point? Can we compare the strategy they use in their uninstlalation process with the one foreseen of NVDA and highlight the differences if any?

I may have missed something. Thanks.

**Comment by feerrenrut** [comment]
@CyrilleB79 I don't think it is a technical limitation, but instead a UX one; trying to prevent one users decision from irreversibly affecting another user. Perhaps other software does this and we shouldn't worry about it. On the other hand, lots of other software leaves (a minimal data footprint) user config behind when it is uninstalled. This is why I suggest changing the location for add-ons. Keep the user config data in user config, and addon's elsewhere.

**Comment by lukaszgo1** [comment]
> On the grander scheme, I'd like to see add-ons stored somewhere shared like app-data rather than the user profile. Then add-ons can be shared by users. This is a separate idea, that we don't need to discuss here, but would alleviate this issue.



No - it would not. While this problem started as a "add-ons are taking a lot of space after NVDA is uninstalled" the more important matter is that users expect to be able to remove NVDA's configh as part of uninstallation and currently this is impossible.
Also the idea of storing add-ons outside of users config is nice in theory but not very good in practice. Many add-ons store their config in their own folder this would be most noticeable for commercial add-ons such as Vocalizer from Code Factory which stores the license information inside its own folder structure. This is problematic since the add-on may be licensed to only a single user of the machine. I also expect this can be problematic if multiple users are logged at the same time.

> Many applications (such as Jaws, Matlab, but maybe also browsers, Notepad++) ask if user config should also be removed or not when uninstalled. I have always imagined that checking the "Also remove user data" caused data removal for all users. But I have never checked. Do you know more on this point? 

In most cases (JAWS, Notepad++) only configuration of a currently logged-in user is removed.

**Comment by feerrenrut** [comment]
> No - it would not. While this problem started as a "add-ons are taking a lot of space after NVDA is uninstalled" the more important matter is that users expect to be able to remove NVDA's config as part of uninstallation and currently this is impossible.

This issue has jumped straight to a solution without first exploring the problem. The "addons take up space after uninstall" is a good example of a problem statement, I think this issue should focus on that. I'm happy to create a separate issue to cover "all user config should be removed on uninstall" however that isn't a good example of a problem statement, it is a solution missing a motivation, I.E. why does a user want this.
Some guesses about 'why':
- To ensure that user config is in a clean state to fix prior misconfiguration.
- To ensure the privacy of ex-users.
We can continue this discussion on the new issue.

Outside of add-ons, user config does not take up much space, and only the most specialized use cases of NVDA would have an issue (IE perhaps some embedded use-case), this is not something we should focus on without specific input from someone working on that.

>Many add-ons store their config in their own folder this would be most noticeable for commercial add-ons such as Vocalizer from Code Factory which stores the license information inside its own folder structure. This is problematic since the add-on may be licensed to only a single user of the machine. I also expect this can be problematic if multiple users are logged at the same time.

Yes, this would be a concern. add-ons would have to start storing their config in a more suitable location. Storing / retrieving config sounds like something NVDA should provide an explicit API for. This may be as simple as providing a path to the addon config data.

>In most cases (JAWS, Notepad++) only configuration of a currently logged-in user is removed.

This is interesting, and a useful data point. Other applications certainly leave user config data by default (including Microsoft apps). I haven't found any in-depth guidance on this from Microsoft, [they do suggest that user config should be removed on uninstall](https://blogs.windows.com/windowsdeveloper/2016/05/10/getting-started-storing-app-data-locally/), [also here](https://docs.microsoft.com/en-us/windows/apps/design/app-settings/store-and-retrieve-app-data#storing-app-data-in-the-app-data-stores).

Some applications can be installed "just for the current user", this may affect the behavior here also.

**Comment by jcsteh** [comment]
> Other applications certainly leave user config data by default (including Microsoft apps). I haven't found any in-depth guidance on this from Microsoft, [they do suggest that user config should be removed on uninstall](https://blogs.windows.com/windowsdeveloper/2016/05/10/getting-started-storing-app-data-locally/), [also here](https://docs.microsoft.com/en-us/windows/apps/design/app-settings/store-and-retrieve-app-data#storing-app-data-in-the-app-data-stores).

This is a fundamental problem with Windows desktop apps and has been thus since the beginning. Both these articles are about apps in containers (UWP and now MSIX), which automatically manage uninstallation of app data across the system. This was one of the major reasons Microsoft implemented MSIX: to solve this problem for desktop apps.

**Comment by lukaszgo1** [comment]
> > Many add-ons store their config in their own folder this would be most noticeable for commercial add-ons such as Vocalizer from Code Factory which stores the license information inside its own folder structure. This is problematic since the add-on may be licensed to only a single user of the machine. I also expect this can be problematic if multiple users are logged at the same time.
> 
> Yes, this would be a concern. add-ons would have to start storing their config in a more suitable location. Storing / retrieving config sounds like something NVDA should provide an explicit API for. This may be as simple as providing a path to the addon config data.


In case of the Code Factory add-on it is certainly not as simple.
The configuration and licencing data for the voices is managed outside of NVDA's UI by a separate process. Since this binary is closed source it is impossible to understand why this is done like that.

**Comment by britechguy** [comment]
Well, this very issue has been discussed on the NVDA User Group this very evening.

To me, it makes sense to offer a user the ability to remove their personal settings at uninstall AND also at install, with the default behavior being to leave them in place in both situations.

It's quite conventional for uninstallers where something has been installed such that all users can use the software to only be able to remove personal settings for the account doing the uninstall.  It's a fact of security without gyrations that I can't see as being warranted in the vast majority of situations.

But it's also quite common for NVDA users to do uninstall and reinstall sequences as part of troubleshooting, and if they've neglected to remove their personal settings at uninstall, those have the potential to be a part of the problem.  So if personal settings are detected at install time, the option should also be given to remove those at that time.

Whether uninstalling or installing on a machine where eariler user settings are detected, the default behavior should be to leave them in place unless the user specifically chooses to remove them.

The removal option is likely to be exercised very seldom, but it's far better to have the installer/uninstaller doing this removal than for an end user to go noodling around trying to do it by hand.

**Comment by CyrilleB79** [comment]
Another example of a person fooled by the config data remaining in the user config: https://github.com/nvaccess/nvda/issues/17834#issuecomment-3790510157.

**Comment by seanbudd** [comment]
This is another good reason for #18450 so we can build as MSIX
